week7/NYTArticles/app/src/main/java/com/example/nytarticles/view/ArticlesFragment.kt

```diff
@@ -0,0 +1,144 @@
+        ViewModelProvider(this).get(ArticlesViewModel::class.java)
+    }
+
+    // Inflate the layout for this fragment
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return inflater.inflate(R.layout.fragment_articles, container, false)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        fetchDataFromInternet()
+        fetchDataFromDB()
+        fab.setOnClickListener {
+            articlesViewModel.getFavoriteArticlesListFromDB()
+                .observe(this.viewLifecycleOwner, Observer {
+                    if (it.isEmpty())
+                        requireContext().toast(getString(R.string.no_favorite_articles))
+                    else {
+                        articlesRV.layoutManager = null
+                        findNavController().navigate(ArticlesFragmentDirections.actionArticlesToFavorites())
+                    }
+
+                })
+        }
+        /**
+         * retrieving the state of the recyclerview
+         */
+        if (savedInstanceState != null) {
+            articlesRV.layoutManager = layoutManager
+            articlesRV.scrollToPosition(savedInstanceState.getInt(ARTICLES_SCROLL_KEY))
+            Log.d(TAG, "position: ${savedInstanceState.getInt(ARTICLES_SCROLL_KEY)}")
+        }
+
+    }
+
+
+    /**
+     * Fetching articles from the Database if the internet is not connected
+     */
+
+    private fun fetchDataFromDB() {
+        articlesViewModel.getArticlesListFromDB().observe(this.viewLifecycleOwner, Observer {
+            Log.d(TAG, "it = $it")
+            setupRecyclerView(it)
+        })
+
+    }
+
+    /**
+     * Fetching articles from the internet
+     */
+    private fun fetchDataFromInternet() {
+        networkStatusChecker.performIfConnectedToInternet {
+            articlesViewModel.getArticleList().observe(this.viewLifecycleOwner, Observer {
+                Log.d(TAG, "internet data = $it")
+                articlesViewModel.saveArticlesListToDB(it)
+            })
+        }
+    }
+
+    /**
+     * attaching the layoutManager and the adapter to the recyclerview
+     */
+    private fun setupRecyclerView(articles: List<Article>) {
+        articlesRV.layoutManager = layoutManager
+        articlesRV.adapter = ArticleAdapter(articles, listener = {
+            if (it is Article) {
+                if (!it.isFavorite) {
+                    it.isFavorite = true
+                    articlesViewModel.saveFavoriteArticleToDB(it)
+                    requireContext().toast(getString(R.string.added_to_favorites))
+                } else {
+                    it.isFavorite = false
+                    articlesViewModel.saveFavoriteArticleToDB(it)
+                    requireContext().toast(getString(R.string.removed_from_favorites))
+                }
+
+            } else {
```

Comment:
Okay this is why you have Any.

maybe have two different listeners then, to have the code a bit more separate?